### ISSUE TITLE
[META] Support PHP7.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1386]
This will initially serve as a place to discuss _when_ (assuming not _if_) we intent to support PHP7. Later on, we can file separate issues and list them here as part of this meta. PHP7 is currently in [RC8](https://secure.php.net/index.php#id2015-11-26-1) with the intention of a final release before the end of the year (initial set date was Nov 12 - now passed).

Drupal officially supports it (respective issue in d.org: [[META] Support PHP 7 [#2454439]](https://www.drupal.org/node/2454439)) and they seem to have adapted their testbot/CI as well (respective issue: [Support PHP7 [#2461603]](https://www.drupal.org/node/2461603)). You might have already noticed tests in the issue queue on d.org being run on various versions of php - including php7.

Zend speak of [25% - 70% performance increase](http://www.zend.com/en/resources/php-7) and in d.org they [mention php7 being ~50% faster than 5.5](https://www.drupal.org/node/2454439#comment-9769093). If that is the case, I expect the adoption rate in providers as well as Linux distros to be rather rapid.

**Comment by klonos** [comment]
...what triggered this issue personally for me is that the stack that I use locally for my tests now officially supports php7 builds for both x86 as well as x64 (on Windows that is). So, I was thinking to start using it. Didn't want to be shot down by "we don't do php7 ...yet" replies in related issues that I file though. So I thought I might ask first. :wink:

**Comment by klonos** [comment]
...also talking about D8 feature parity concerns, in my search for more benchmarks specifically with Drupal 7 (because its users are basically our main target audience - the ones not wanting to upgrade to D8 that is) and php7, I came across this blog: [Benchmarking Drupal 7 on PHP 7-dev | Drupal On Windows](http://www.drupalonwindows.com/en/blog/benchmarking-drupal-7-php-7-dev)

Besides the performance tests (58% increase!, plus another 16% with page caching enabled!!!), the comments section was intriguing (though only two comments):

> Thanks for sharing this valuable information. I still want to ask if Drupal 7 (and major contributed modules) are compatible with PHP-7 ? What are your experiences regarding this? Thanks.

reply:

> There is a big chance that we will never see D7 running on PHP7 because D7 is "abandoned". Just take a look at this [issue](https://www.drupal.org/node/2297817). Basically no one cares about D7 anymore because all the small sites have fled to Wordpress and the big ones cannot keep on working with the shortcomings of D7 - so they are focusing only on D8.

Story sounds familiar, no?

The mentioned issues is [Do not attempt field storage write when field content did not change [#2297817]](https://www.drupal.org/node/2297817) which was *critical* and ended up having its own issue for D7: [[Drupal 7] Do not attempt field storage write when field content did not change [#2470619]](https://www.drupal.org/node/2470619). In the D7 issue, you see this in the issue summary (emphasis is mine):

> Issue originally opened at #2297817: ... and **kept hostage for a year because of Drupal 8**. Now that the Drupal 8 patch does not seems as obvious as it was and discussion is going further, I open a Drupal-7 only patch as per the backport policy states (actually it should be the other way around but the original issue is now too deep in Drupal 8 to change back).

**Comment by klonos** [comment]
...all this "smells" like momentum opportunity to me. :smiling_imp:

**Comment by quicksketch** [comment]
Hm, from that same blog post: http://www.drupalonwindows.com/en/blog/benchmarking-drupal-7-php-7-dev

The author mentions:

> This is a 100% default D7 setup on Windows (IIS 8.5) against a local MySQL database and no modifications whatsoever with Zend Opcache turned on.

So it sounds like at least for him, D7 ran well enough to do the benchmarks with no changes. In any case, I think you're absolutely right. Backdrop needs to support PHP7. If it doesn't already, we should plan on supporting it right away. We can [enable additional PHP version checking on Travis CI](https://github.com/travis-ci/travis-ci/issues/2480#issuecomment-75988205) with just a single change. I'll give that a go and see what sort of errors crop up. This will run our entire test suite against multiple versions of PHP. I'd held off on that b/c I didn't want to suck up too much of Travis's resources, but the differences between PHP 5.3 (what we're using now, our minimum version) and PHP 7 may be so great that it is required that we test it separately.

**Comment by quicksketch** [comment]
Filed a PR at https://github.com/backdrop/backdrop/pull/1128 just to see what would happen. It should spin up 2 separate Travis CI servers and test them in parallel, one on PHP 5.3 and another on PHP 7. Though since 7.0 isn't scheduled to be released until tomorrow(!) I'm not sure if my current line will work or not.

**Comment by quicksketch** [comment]
Of course it's not just so easy as adding a single line. :stuck_out_tongue: 

Looks like the default php-fpm.conf file is different in PHP 7, so copying the existing file won't work consistently from PHP 5.3 to PHP 7. That's okay, we can copy the relevant parts into our own default php-fpm.conf file and ship it with Backdrop in the /core/etc/travis-ci folder. We already do this for the Apache configuration file.

A few more attempts at the PR and I think we should be able to get it going.

**Comment by quicksketch** [comment]
Yay, tests are now running (in the PR) on PHP 7!

Lots of failures, but 99% of the failures are one of the two of these:
```
Exception Deprecated bootstrap.inc     3520 require_once()                     
    Methods with the same name as their class will not be constructors in a
    future version of PHP; views_many_to_one_helper has a deprecated constructor

Exception Warning    user.pages.inc     213 user_logout()                      
    session_destroy(): Session callback expects true/false return value
```

I'll update the PR to fix these two trivial problems and see how far it gets.

**Comment by biolithic** [comment]
Yea!

**Comment by quicksketch** [comment]
Yay, not a bad test run! 15 total failing tests on PHP 7: https://travis-ci.org/backdrop/backdrop/builds/94546986

PHP 7 is *definitely* a lot faster! The tests finished in 10min, 42 sec. The PHP 5.3 tests took 18min, 14sec! Of course the PHP 7 tests may have just gotten a faster test-runner, but still I think we can expect PHP 7 to win out most of the time.

**Comment by serundeputy** [comment]
:bow: :+1:

**Comment by quicksketch** [comment]
PR updated again, now we just have two failing tests. One in GD related to image rotation (an open issue at https://www.drupal.org/node/2215369) and one related to Admin Bar which I can't reproduce locally.

Current results: https://travis-ci.org/backdrop/backdrop/jobs/94585202

But hey, that's really close! I also set up a local PHP 7 RC8 and tried out the basics (Views, Content Types, Fields, Layouts, etc). Basically exactly as you would expect, no new errors. Installation works, which is key because the new opcode cache by default sets to not refresh files off disk when they are edited, they have to be manually flushed (which we were doing already).

**Comment by quicksketch** [comment]
I reviewed the Drupal issue for PHP7, but none of the bugs actually applied to Backdrop (mostly dealing with class names that we don't have in Backdrop). There were a number of PHP 5.5+ issues that I had to bundle into the current PR, but most of those seem as though they haven't made it back to D7 yet, which is why we haven't cross-ported them to Backdrop.

The two unresolved D8 issues both apply to Backdrop but appear to be environment-specific fixes:

- https://www.drupal.org/node/2455455
- https://www.drupal.org/node/2455465

**Comment by klonos** [comment]
Thanx for the quick response. It's great news that we are mostly ok. I'll be testing on actual test environments soon and we'll see if what other issues might surface.

**Comment by quicksketch** [comment]
We now have 100% tests passing on PHP 7 :smile:

https://github.com/backdrop/backdrop/pull/1128

I figure I better explain each fix one at a time, as there are a lot of seemingly unrelated changes in here.

```diff
-  - sudo cp $HOME/.phpenv/versions/`php -r "print PHP_VERSION;"`/etc/php-fpm.conf.default $HOME/.phpenv/versions/`php -r "print PHP_VERSION;"`/etc/php-fpm.conf
-  - $HOME/.phpenv/versions/`php -r "print PHP_VERSION;"`/sbin/php-fpm
+  - sudo cp core/misc/travis-ci/php-fpm.conf $HOME/.phpenv/versions/$(phpenv version-name)/etc/php-fpm.conf
+  - $HOME/.phpenv/versions/$(phpenv version-name)/sbin/php-fpm
```

This change to Travis CI is to accommodate for the location of Travis CI installing PHP not necessarily matching the PHP version number. It had previously... but in the case of PHP 7, the version number is "7.0.1-dev" but the directory name is "7.0-nightly". Using the `$(phpenv version-name)` is the most accurate way of getting the correct directory name rather than getting a version number from PHP.

The new `core/misc/travis-ci/php-fpm.conf` should provide us with more consistent (and faster) Travis CI runtimes. Because the php-fpm.conf file included in PHP 5.3 and 7 differs, including our own that is the same for both provides greater consistency. It also allows us to set up enough PHP processes to match our number of concurrent tests. Previously we running 12 concurrent tests on only 5 PHP processes. Now we start 16 processes that can be shared between the concurrent tests.

```diff
+  // Standardize $degrees to be an integer if possible. This avoids strange
+  // rounding errors when rotating multiples of 90 degrees. For example a
+  // 100x100 image when rotated 90.0 (float) degrees may result in an image
+  // dimension of 99x100 pixels. See https://bugs.php.net/bug.php?id=65148.
+  if (strcmp((int) $degrees, $degrees) === 0) {
+    $degrees = (int) $degrees;
+  }
+  else {
+    $degrees = (float) $degrees;
+  }
+  // Likewise, negative rotations can sometimes have rounding errors, convert
+  // to a positive number before passing to the toolkit. This bug is also
+  // reported at https://bugs.php.net/bug.php?id=65148 (same as above).
+  while ($degrees < 0) {
+    $degrees = 360 + $degrees;
+  }
+  // Reduce greater than 360 degrees just for good measure.
+  while ($degrees >= 360) {
+    $degrees = 360 - $degrees;
+  }
```

PHP has special-cased rotation if at 90, 180, or 270 degrees to ensure the dimensions stay the same. Oddly, it hasn't special cased -90 degrees or 450, so we have to normalize these values before passing to GD (i.e -90 degrees is "fixed" to 270 degrees). There is a PHP bug at https://bugs.php.net/bug.php?id=65148 for this and a Drupal issue at https://www.drupal.org/node/2215369.

All changes in `session.inc` (mostly just returning `TRUE` everywhere) is required in PHP 7. Without it, sessions throw errors all over the place. Any session handler MUST return either TRUE or FALSE. Returning NULL is not acceptable.

```diff
-    node_type_delete('special');
-    menu_rebuild();
+    $this->backdropPost('admin/structure/types/manage/special/delete', array(), t('Delete'));
+    $this->backdropGet('js/admin_bar/cache/' . $hash);
 
-    $this->backdropGet('');
```

This fix to Admin Bar actually was a bug. We weren't checking the admin bar cache previously (where the actual admin bar is displayed). Previously this test passed coincidentally.

In Date's `date.test`:
```diff
       // Move the date right below the title.
-      'weight' => -4,
+      'weight' => -5,
```

We did this in a number of places. Date tried to position the new Date field below the node title field. However, the "Body" field also had a weight of -5. In PHP 5.3, that Date field came first. In PHP 7, the Body field came first. As PHP has always said that comparing two identical values will result in an undefined order on output, it looks like that finally caused a problem. Now with -5 weight, the date field will always come first.

In `filter.test`:
```diff
-    $f = filter_xss('<img src=" &#14;  javascript:alert(0)">', array('img'));
+    // Note that with &#14; this fails in PHP 5.4+. Work-around solutions are
+    // being discussed in https://www.drupal.org/node/1210798.
+    $f = filter_xss('<img src=" &#x09;  javascript:alert(0)">', array('img'));
```

This issue arose in PHP 5.4 apparently. Drupal has applied a similar temporary fix, though it's unclear if this XSS precaution is actually exploitable in modern browsers. It appears to be an IE6 fix that is generally no longer an issue. Drupal is discussing more comprehensive solutions at the issue mentioned in the comment: https://www.drupal.org/node/1210798

```diff
-function image_style_revert_form($form, $form_state, $style) {
+function image_style_revert_form($form, &$form_state, $style) {
   $form_state['image_style'] = $style;
```

In a couple of places we weren't accepting $form_state by reference but we were modifying it in a way we expected it to be saved/passed into the validate and submit handlers. I'm not sure how this worked in PHP 5.3, but it's nice that PHP 7 pointed out that we were missing the `&` to pass by reference.

In `common_test.module`:
```diff
-  return t('This tests the inline CSS!') . backdrop_add_css($css, 'inline');
+  backdrop_add_css($css, 'inline');
+  return 'This tests the inline CSS!';
```

PHP 7 now throws a warning when concatenating a string to an array. Why were we doing this in the first place? No idea.

In `image.test`:

```diff
+        // PHP 5.5 GD bug: https://bugs.php.net/bug.php?id=65148. PHP 5.5 GD
+        // rotates differently then it did in PHP 5.4 resulting in different
+        // dimensions. For the test images, the dimensions will be 1 pixel
+        // smaller in both dimensions (though other tests have shown a
+        // difference of 0 to 3 pixels in both dimensions.
+        // @todo: Add an upper limit version check if this is fixed in PHP.
+        if (version_compare(PHP_VERSION, '5.5', '>=') && $values['function'] === 'rotate' && $values['arguments'][0] % 90 != 0) {
+          $values['height']--;
+          $values['width']--;
+        }
```

This is probably the only change I'm not really satisfied with. This came from the Drupal issue at https://www.drupal.org/node/2215369, where they have applied this temporary fix to D8 and is pending for Drupal 7. In short, PHP 5.5+ has a bug in it that causes it to incorrectly trim off pixels when rotating images. PHP 5.3 and 5.4 rotate properly. If this is fixed in https://bugs.php.net/bug.php?id=65148, we will need to update our tests at some point.

In `image.gd.inc`:
```php
-    // If no transparent colors, use white.
-    if ($background == 0) {
+    // If using a PNG with alpha transparency, the color may be -1. Allocate a
+    // dedicated full alpha transparent color.
+    if ($background === -1) {
+      $background = imagecolorallocatealpha($image->resource, 255, 255, 255, 127);
+    }
+    // Allocating a transparent color may fail (e.g. on a jpg). Use white as
+    // a fallback.
+    if ($background === FALSE) {
       $background = imagecolorallocatealpha($image->resource, 255, 255, 255, 0);
     }
```

In addition to returning FALSE, PHP may now return -1 when detecting if an image has a transparent color defined. It returns FALSE if there is no transparency defined. And returns -1 if it's a 24-bit PNG that supports transparency (apparently).

```diff
-  return (function_exists($function) && $image->resource = $function($image->source));
+  if (function_exists($function)) {
+    $image->resource = $function($image->source);
+  }
+  return !empty($image->resource);
```

These sort of IF statements where assignment is involved apparently does not work well in PHP 7. That's fine with me, as the order of operations has always been unclear to me anyway. I think this could have been solved with additional parenthesis, but I prefer to separate assignment from IF statements anyway.

In `system.test`:
```diff
+  private $assertErrors = FALSE;
```

PHP now throws errors if you try to access a property that has not be defined, e.g. `$this->assertErrors`, so we now define this property before using it.

In `views/includes/handlers.inc`:
```diff
-  function views_many_to_one_helper(&$handler) {
+  function __construct(&$handler) {
     $this->handler = &$handler;
   }
```

PHP 7 removed the old-style constructors (it's about time) where the class just had a function with the same name that it called as a constructor.

But that's all she wrote. With that set of changes we're fully passing tests. There are probably smaller follow-up issues (like those Drupal issues I posted in my last comment.

**Comment by docwilmot** [comment]
Well done, thanks for the detailed explanation.

**Comment by klonos** [comment]
> We now have 100% tests passing on PHP 7 :smile:

Woot-woot :tada: Thanx @quicksketch! Do you want me to file respective placeholder issues for those d.org issues you mentioned? ...or is "following" the D7/D8 changelog sufficient for us?

**Comment by klonos** [comment]
@sutibun: Another thing to list as a Backdrop feature I guess: "PHP7 support for improved performance! Up to 75% faster!!!" ...OK, the speed improvement itself is not a Backdrop feature - it comes from PHP7, but we are talking marketing in https://github.com/backdrop-ops/backdropcms.org/issues/122 and we all know that marketing people have sold their souls :smiling_imp:

**Comment by quicksketch** [comment]
> Do you want me to file respective placeholder issues for those d.org issues you mentioned? ...or is "following" the D7/D8 changelog sufficient for us?

Actually it looks like those drupal.org issues are for PHP 5.6+ (including 7), but generally don't seem to be needed and have been causing some side-effects. Let's wait until some solution is settled for D7 and do the cross-port when it's ready.

**Comment by quicksketch** [comment]
I've merged this into 1.x and 1.2.x, so it will be in the next bug-fix release (1.2.3).

**Comment by docwilmot** [comment]
Note that I just pushed a PR which ran slower on PHP 7 than on 5. Did something break?
https://travis-ci.org/docwilmot/backdrop/builds/95573836

**Comment by quicksketch** [comment]
> Note that I just pushed a PR which ran slower on PHP 7 than on 5. Did something break?

No, I've seen test runners come in similar and sometimes drastically different. It really just depends on the speed of the server that runs the test. Though in general PHP7 should be faster, it won't always be the case. We'd need to do actual benchmarks on the same machine to determine the real difference.

**Comment by MrHaroldA** [comment]
FYI: Arch Linux pushed php 7.0.1 to the extra packages this weekend, so a lot of devvers running Arch will chip in on the php7 support soon!
